readme.md
# ControlMap (De)Serializer Library

The controlmap.txt parser for Skyrim.

## Table of Contents

- [ControlMap (De)Serializer Library](#controlmap-deserializer-library)
  - [Table of Contents](#table-of-contents)
  - [Features](#features)
- [Examples](#examples)
  - [License](#license)

## Features

- [x] `controlmap.txt` => `json`
- [x] `json` => `controlmap.txt` (formatted with preserved comments)
- [x] Analysis using enum scanCodes.

# Examples

- PreRequirements

```toml
[dependencies]
controlmap_parser = "*"
serde_json = "1.0" # Json converter
```

```rust
use controlmap_parser::ControlMap;

type Result<T, E = Box<dyn std::error::Error + 'static>> = core::result::Result<T, E>;
fn main() -> Result<()> {
    let input = r#"
// Lockpicking
RotatePick	0xff		0xa		0x000b	0	0	0	0x8
RotateLock	0x1e,0xff	0xff	0x000c	0	0	0	0x8
DebugMode	0x35		0xff	0x4000	0	0	0	0x8
Cancel		0x0f		0xff	0x1000	0	0	0	0x8

// Favor
Cancel	0x0f	0xff	0x1000	0	0	0	0x108"#;

    let control_map = ControlMap::from_txt(input)?;
    println!("txt => Struct:\n{:?}", &control_map);


    // txt => json
    let expected_json = r#"{
  "lines": [
    "BlankLine",
    {
      "Comment": " Lockpicking"
    },
    {
      "EventLine": {
        "event_name": "RotatePick",
        "keyboard_id": {
          "One": "0xff"
        },
        "mouse_id": {
          "One": "0xa"
        },
        "gamepad_id": {
          "One": "0x000b"
        },
        "remap_key": false,
        "remap_mouse": false,
        "remap_gamepad": false,
        "event_binary_flag": "0x8"
      }
    },
    {
      "EventLine": {
        "event_name": "RotateLock",
        "keyboard_id": {
          "Or": [
            {
              "One": "0x1e"
            },
            {
              "One": "0xff"
            }
          ]
        },
        "mouse_id": {
          "One": "0xff"
        },
        "gamepad_id": {
          "One": "0x000c"
        },
        "remap_key": false,
        "remap_mouse": false,
        "remap_gamepad": false,
        "event_binary_flag": "0x8"
      }
    },
    {
      "EventLine": {
        "event_name": "DebugMode",
        "keyboard_id": {
          "One": "0x35"
        },
        "mouse_id": {
          "One": "0xff"
        },
        "gamepad_id": {
          "One": "0x4000"
        },
        "remap_key": false,
        "remap_mouse": false,
        "remap_gamepad": false,
        "event_binary_flag": "0x8"
      }
    },
    {
      "EventLine": {
        "event_name": "Cancel",
        "keyboard_id": {
          "One": "0x0f"
        },
        "mouse_id": {
          "One": "0xff"
        },
        "gamepad_id": {
          "One": "0x1000"
        },
        "remap_key": false,
        "remap_mouse": false,
        "remap_gamepad": false,
        "event_binary_flag": "0x8"
      }
    },
    "BlankLine",
    {
      "Comment": " Favor"
    },
    {
      "EventLine": {
        "event_name": "Cancel",
        "keyboard_id": {
          "One": "0x0f"
        },
        "mouse_id": {
          "One": "0xff"
        },
        "gamepad_id": {
          "One": "0x1000"
        },
        "remap_key": false,
        "remap_mouse": false,
        "remap_gamepad": false,
        "event_binary_flag": "0x108"
      }
    }
  ]
}"#;
    let json = serde_json::to_string_pretty(&control_map)?;
    println!("struct => json:\n{:?}", &control_map);
    assert_eq!(&json, expected_json);

    let re_control_map: ControlMap = serde_json::from_str(&json)?;
    println!("json => txt:\n{}", &re_control_map);
    // Yes, blank lines are not removed, they are applied as is.
    // There are places where it would be an error to do otherwise.
    let formatted_control_map = r#"
// Lockpicking
RotatePick	0xff	0xa	0x000b	0	0	0	0x8
RotateLock	0x1e+0xff	0xff	0x000c	0	0	0	0x8
DebugMode	0x35	0xff	0x4000	0	0	0	0x8
Cancel	0x0f	0xff	0x1000	0	0	0	0x8

// Favor
Cancel	0x0f	0xff	0x1000	0	0	0	0x108
"#;
    assert_eq!(re_control_map.to_string(), formatted_control_map);

    Ok(())
}
```

```shell
cargo run --example cast_and_fmt --features serde
cargo run --example parse_and_print --features serde
cargo run --example scan_code
```

## License

[MIT](https://opensource.org/licenses/MIT) or
[Apache 2.0](https://opensource.org/license/apache-2-0/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2023-11-17
### :boom: BREAKING CHANGES
- due to [`4bd3781`](https://github.com/SARDONYX-sard/controlmap-parser/commit/4bd3781a5f6c143f3f71b0a103ccfbc54a2157e9) - create new parser by nom *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*:

  create new parser by nom


### :sparkles: New Features
- [`4bd3781`](https://github.com/SARDONYX-sard/controlmap-parser/commit/4bd3781a5f6c143f3f71b0a103ccfbc54a2157e9) - create new parser by nom *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*
- [`76bc0b3`](https://github.com/SARDONYX-sard/controlmap-parser/commit/76bc0b35f0945787dd525fed5911f8ff73d5fb72) - **github**: add template issue *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*
- [`2565fba`](https://github.com/SARDONYX-sard/controlmap-parser/commit/2565fba6a36b9f32ff1bbe14dee4ee98331b50fd) - **serde**: support json to txt *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*
- [`c38d352`](https://github.com/SARDONYX-sard/controlmap-parser/commit/c38d3521d6dfe10a1579a1d58aa6dbc5dd607691) - **examples**: add examples *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*

### :bug: Bug Fixes
- [`b60877e`](https://github.com/SARDONYX-sard/controlmap-parser/commit/b60877e478f216cb10cf5e56157bdea3f8adccff) - fix ci *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*
- [`ea3bcf1`](https://github.com/SARDONYX-sard/controlmap-parser/commit/ea3bcf1575a38b469a15e01e80004b8437d4041d) - **make**: fix make run name *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*
- [`aa8772e`](https://github.com/SARDONYX-sard/controlmap-parser/commit/aa8772e92c83d66346558495b6ef0b81d98265b4) - **make**: add phony *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*
- [`6538902`](https://github.com/SARDONYX-sard/controlmap-parser/commit/6538902fc9935000f3641b658bd2e37b0817ee48) - **ci**: fix tag name *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*

### :wrench: Chores
- [`a43e40e`](https://github.com/SARDONYX-sard/controlmap-parser/commit/a43e40e8fc1e897116899f9376dbb7cd72a5e792) - **license**: add files *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*
- [`c15402f`](https://github.com/SARDONYX-sard/controlmap-parser/commit/c15402fe100d39d88a5434ede54a18e35b6e51aa) - change settings *(commit by [@SARDONYX-sard](https://github.com/SARDONYX-sard))*


[0.3.0]: https://github.com/SARDONYX-sard/controlmap-parser/compare/0.2.0...0.3.0

src/lib.rs
#![allow(clippy::tabs_in_doc_comments)] // <- controlmap.txt ignores clippy's caution because tab is part of the language use.
//! # ControlMap Parser Library
//!
//! The controlmap.txt parser for Skyrim.
//!
//! ## features
//!
//! - [x] controlmap.txt => json structure
//! - [x] json structure => controlmap.txt
//! - [x] Analysis using enum scanCodes.
//!
//! # Examples
//!
//! ```rust
//! use controlmap_parser::ControlMap;
//!
//! type Result<T, E = Box<dyn std::error::Error + 'static>> = core::result::Result<T, E>;
//! fn main() -> Result<()> {
//!     let input = r#"
//! // Lockpicking
//! RotatePick	0xff		0xa		0x000b	0	0	0	0x8
//! RotateLock	0x1e,0xff	0xff	0x000c	0	0	0	0x8
//! DebugMode	0x35		0xff	0x4000	0	0	0	0x8
//! Cancel		0x0f		0xff	0x1000	0	0	0	0x8
//!
//! // Favor
//! Cancel	0x0f	0xff	0x1000	0	0	0	0x108"#;
//!
//!     let control_map = ControlMap::from_txt(input)?;
//!     println!("txt => Struct:\n{:?}", &control_map);
//!
//!     let json = serde_json::to_string_pretty(&control_map)?;
//!     let expected_json = r#"{
//!   "lines": [
//!     "BlankLine",
//!     {
//!       "Comment": " Lockpicking"
//!     },
//!     {
//!       "EventLine": {
//!         "event_name": "RotatePick",
//!         "keyboard_id": {
//!           "One": "0xff"
//!         },
//!         "mouse_id": {
//!           "One": "0xa"
//!         },
//!         "gamepad_id": {
//!           "One": "0x000b"
//!         },
//!         "remap_key": false,
//!         "remap_mouse": false,
//!         "remap_gamepad": false,
//!         "event_binary_flag": "0x8"
//!       }
//!     },
//!     {
//!       "EventLine": {
//!         "event_name": "RotateLock",
//!         "keyboard_id": {
//!           "Or": [
//!             {
//!               "One": "0x1e"
//!             },
//!             {
//!               "One": "0xff"
//!             }
//!           ]
//!         },
//!         "mouse_id": {
//!           "One": "0xff"
//!         },
//!         "gamepad_id": {
//!           "One": "0x000c"
//!         },
//!         "remap_key": false,
//!         "remap_mouse": false,
//!         "remap_gamepad": false,
//!         "event_binary_flag": "0x8"
//!       }
//!     },
//!     {
//!       "EventLine": {
//!         "event_name": "DebugMode",
//!         "keyboard_id": {
//!           "One": "0x35"
//!         },
//!         "mouse_id": {
//!           "One": "0xff"
//!         },
//!         "gamepad_id": {
//!           "One": "0x4000"
//!         },
//!         "remap_key": false,
//!         "remap_mouse": false,
//!         "remap_gamepad": false,
//!         "event_binary_flag": "0x8"
//!       }
//!     },
//!     {
//!       "EventLine": {
//!         "event_name": "Cancel",
//!         "keyboard_id": {
//!           "One": "0x0f"
//!         },
//!         "mouse_id": {
//!           "One": "0xff"
//!         },
//!         "gamepad_id": {
//!           "One": "0x1000"
//!         },
//!         "remap_key": false,
//!         "remap_mouse": false,
//!         "remap_gamepad": false,
//!         "event_binary_flag": "0x8"
//!       }
//!     },
//!     "BlankLine",
//!     {
//!       "Comment": " Favor"
//!     },
//!     {
//!       "EventLine": {
//!         "event_name": "Cancel",
//!         "keyboard_id": {
//!           "One": "0x0f"
//!         },
//!         "mouse_id": {
//!           "One": "0xff"
//!         },
//!         "gamepad_id": {
//!           "One": "0x1000"
//!         },
//!         "remap_key": false,
//!         "remap_mouse": false,
//!         "remap_gamepad": false,
//!         "event_binary_flag": "0x108"
//!       }
//!     }
//!   ]
//! }"#;
//!     assert_eq!(&json, expected_json);
//!
//!     let control_map: ControlMap = serde_json::from_str(&json)?;
//!     // Yes, blank lines are not removed, they are applied as is.
//!     // There are places where it would be an error to do otherwise.
//!     let formatted_control_map = r#"
//! // Lockpicking
//! RotatePick	0xff	0xa	0x000b	0	0	0	0x8
//! RotateLock	0x1e+0xff	0xff	0x000c	0	0	0	0x8
//! DebugMode	0x35	0xff	0x4000	0	0	0	0x8
//! Cancel	0x0f	0xff	0x1000	0	0	0	0x8
//!
//! // Favor
//! Cancel	0x0f	0xff	0x1000	0	0	0	0x108
//! "#;
//!     assert_eq!(control_map.to_string(), formatted_control_map);
//!
//!     Ok(())
//! }
//! ```

mod controlmap;
pub mod parser;
pub mod scan_code;

pub use controlmap::{ControlMap, ControlMapError};
